Add apitest rolling offer simulation script

ghubstan opened the pull request:
The script demonstrates a way to always keep one offer in the market.  As soon as the offer is taken and the trade is completed, another offer is placed.

PR https://github.com/bisq-network/bisq/pull/5128 should be reviewed and merge before this one.

ghubstan committed:
Fix bash syntax inconsistencies

Many unnecesary braces around ${VARIABLE} references were removed.

ghubstan committed:
Refactor createpaymentacct and trade protocol functions

Code duplication needs to be reduced as new scripts are added.

- Refactor the createpaymentacct functions.
  The steps required to create a payment account are still
  displayed, but only for Alice, not Bob.

- Treat the trade protocol simulation as an atomic function.
  This will reduce 'main' simulation script size as new
  ones are added.

ghubstan committed:
Move createoffer command generation to its own function

ghubstan committed:
Add apitest rolling offer simulation script

The script demonstrats a way to always keep one offer in the market.
As soon as the offer is taken and the trade is completed, another
offer is placed.

ghubstan committed:
Try alternative syntax to make codacy happy

The codacy issue "$/${} is unnecessary on arithmetic variables"
seems to not apply to $(echo $((1 + $RANDOM % 10))), but we
try another syntax to make her happy.

ghubstan committed:
Add functions for checking trade status

ghubstan committed:
Make executetrade simulation more realistic

Demonstrate how traders poll trade status as they go through
the protocol.

ghubstan committed:
Remove DONE flag

There should not be a check on that flag in the rolling offers script.
Terminate the script with ^C.

sqrrm commented on apitest/scripts/trade-simulation-utils.sh:
```diff
@@ -221,6 +221,271 @@ getfirstofferid() {
+        IS_TRADE_PAYMENT_SENT=$(istradepaymentreceived "$TRADE_DETAIL" "$MAKER_OR_TAKER")
+        exitoncommandalert $?
+        printdate "$SELLER:  Has buyer's payment been transferred to seller's fiat account?  $IS_TRADE_PAYMENT_SENT"
+        if [ "$IS_TRADE_PAYMENT_SENT" = "YES" ]
```
Perhaps a comment to indicate that receipt is assumed as soon as payment sent message is received.

ghubstan replied:
Comment added in commit 072364b4902370fc1a3757640bec896917955f2d.

sqrrm commented on apitest/scripts/rolling-offer-simulation.sh:
```diff
@@ -0,0 +1,119 @@
+#           bitcoind, seednode, arbdaemon, alicedaemon, bobdaemon
+#
+#    These should be run using the apitest harness.  From the root project dir, run:
+#    `$ ./bisq-apitest --apiPassword=xyz --supportingApps=bitcoind,seednode,arbdaemon,alicedaemon,bobdaemon --shutdownAfterTests=false`
```
This kind of simulation would be nice to have for normal regtest that I would run. Could you add instructions on how to run this for my normal alice client without starting the other as I already have them started? I just want a script that uses my normal regtest alice data dir which is already funded and has a history of trades, failed trades and other weird history.

ghubstan replied:
This is related to https://github.com/bisq-network/bisq/pull/5129#discussion_r569491295.   We had a chat on keybase and agreed that dev-test script or java testcases will be added, and these scripts' scope will be limited to serving as non-interactive  demos / tutorials.

sqrrm commented on apitest/scripts/rolling-offer-simulation.sh:
```diff
@@ -0,0 +1,119 @@
+
+    # Generate some btc blocks.
+    printdate "Generating btc blocks after publishing Alice's offer."
+    genbtcblocks 3 2
```
If this is to be used for dev testing I would need to be able to control the block generation completely rather than have them automatically generated.

ghubstan replied:
These simulation scripts are intended to demonstration how the CLI is used to create offers and perform trades.  

A test case that automate one side of a trade, and require manual CLI use from the other side, has been written, but not checked in.  We should talk about the requirements of that test case before I return to work on it and submit it in a new PR.

ghubstan committed:
Explain 'payment sent' implies 'funds deposited'

There is no need to simulate the time it takes between
receiving a 'payment sent' msg and having funds (fiat) deposited
in the seller's bank account.  But there is need to explain
that in the script.

sqrrm reviewed (approved):
utACK
